patches/server/0918-Add-damage-source-wrapper.patch

```diff
@@ -0,0 +1,826 @@
++import io.papermc.paper.entity.damage.wrapper.CustomDamageSourceWrapper;
++import io.papermc.paper.entity.damage.wrapper.CustomEntityDamageSourceWrapper;
++import io.papermc.paper.entity.damage.wrapper.CustomIndirectEntityDamageSourceWrapper;
++import net.kyori.adventure.text.Component;
++import net.minecraft.world.damagesource.DamageSource;
++import net.minecraft.world.damagesource.EntityDamageSource;
++import net.minecraft.world.damagesource.IndirectEntityDamageSource;
++import org.bukkit.craftbukkit.entity.*;
++import org.bukkit.entity.*;
++import org.jetbrains.annotations.NotNull;
++import org.jetbrains.annotations.Nullable;
++
++import java.util.concurrent.ExecutionException;
++import java.util.concurrent.TimeUnit;
++
++public class DamageOriginImpl implements DamageOrigin {
++
++    protected DamageSource origin;
++
++    public DamageOriginImpl(DamageSource origin) {
++        this.origin = origin;
++    }
++
++    public static DamageOrigin fromNMS(DamageSource source) {
++        return new DamageOriginImpl(source);
++    }
++
++    public static DamageOrigin fromAllTypeNMS(DamageSource source) {
++        if(source instanceof EntityDamageSource) {
++            if(source instanceof IndirectEntityDamageSource) {
++                return new IndirectEntityDamageOriginImpl((IndirectEntityDamageSource) source);
++            }
++            return new EntityDamageOriginImpl((EntityDamageSource) source);
++        }
++        return fromNMS(source);
++    }
++
++    @Override
++    public boolean isSweep() {
++        return origin.isSweep();
++    }
++
++    @Override
++    public boolean isCritical() {
++        return origin.isCritical();
++    }
++
++    @Override
++    public boolean isFromProjectile() {
++        return origin.isProjectile();
++    }
++
++    @Override
++    public boolean isFromExplosion() {
++        return origin.isExplosion();
++    }
++
++    @Override
++    public boolean isFromFire() {
++        return origin.isFire();
++    }
++
++    @Override
++    public boolean isFromMagic() {
++        return origin.isMagic();
++    }
++
++    @Override
++    public boolean isFromFalling() {
++        return origin.isFall();
++    }
++
++    @Override
++    public boolean willNotAggro() {
++        return origin.isNoAggro();
++    }
++
++    @Override
++    public boolean willScalesWithDifficulty() {
++        return origin.scalesWithDifficulty();
++    }
++
++    @Override
++    public boolean canBypass(@NotNull DamageBypass damageType) {
++        return switch (damageType) {
++            case ARMOR -> origin.isBypassArmor();
++            case MAGIC -> origin.isBypassMagic();
++            case ENCHANTMENT -> origin.isBypassEnchantments();
++            case INVULNERABILITY -> origin.isBypassInvul();
++        };
++    }
++
++    @Override
++    public boolean applyArmorExhaustion() {
++        return false;
++    }
++
++    @Override
++    public boolean applyMagicExhaustion() {
++        return false;
++    }
```

Comment:
Aren't these supposed to be if it bypasses armor/magic? I'm confused about what these are supposed to represent.

If this is part of the damage bypass things i'd much rather just have these be normal getters.